src/Interpreters/Cache/FileCache.cpp

```diff
@@ -185,70 +187,83 @@ FileSegments FileCache::getImpl(
     /// Given range = [left, right] and non-overlapping ordered set of file segments,
     /// find list [segment1, ..., segmentN] of segments which intersect with given range.
 
-    auto it = files.find(key);
-    if (it == files.end())
-        return {};
+    FileSegments result;
 
-    const auto & file_segments = it->second;
-    if (file_segments.empty())
+    if (enable_limit_download_cache_size && (range.size() > max_enable_download_cache_size))
```

Comment:
1. **max_query_cache_size** and **max_size** are used to limit the overall usage of the cache, and **max_enable_download_cache_size** is to decide whether to cache or not to access the cache for a query.
2. As we mentioned before, in many cases the performance of reading data from the local cache is not as good as reading directly from S3 (even if it already exists in the local cache), because some large-grained caches read from the local cache will suffer from bandwidth limit, rather than reading directly from the higher bandwidth S3 (this is also verified by the results of our tests).

Reply:
Ok, I understood what you mean.

btw one point, for queries like `select * from ... limit n`, the requested range will be the whole file size (because the limit does not affect the requested range unfortunately), e.g. the actually read size is less than the requested, but with your changes it will read from s3 even if there is cached data which could be read from cache completely, e.g. it will not work as you expect if in `limit n`: `n < max_enable_download_cache_size`.

Reply:
Yes, this kind of request is likely to load a lot of data, I think loading them from S3 will perform better than loading from EBS or local SSD. After all, in many scenarios, the access bandwidth provided by S3 far exceeds that of EBS or local physical disks.

Reply:
> Yes, this kind of request is likely to load a lot of data

No, on the contrary, may be you misunderstood the case which I explained, what I meant is -- with this kind of request, the requested range (the boundary of which is limited by 'setReadUntilPosition' method) is the whole file size, but in fact it will stop reading once `limit n` is satisfied, e.g could be much less than it initially assumed it would want to read, but your check makes assumptions based on assumed size. E.g. for example, the worst case, with `limit 1` we will read from s3 with your optimisation (because `limit 1` does not affect the size which is assumed to be read before reading begins), even though this tiny part could be read from cache.

Reply:
OK, I see what you mean. It seems so, because we can't get more upper-level information at the cache layer (such as the limit you mentioned). It seems difficult to solve this problem, unless we request cache relevant information to make judgments